server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -57,12 +60,54 @@
             Response extends ReplicationResponse & WriteResponse
         > extends TransportReplicationAction<Request, ReplicaRequest, Response> {
 
+    private final boolean forceExecutionOnPrimary;
+    private final WriteMemoryLimits writeMemoryLimits;
+    private final String executor;
+
     protected TransportWriteAction(Settings settings, String actionName, TransportService transportService,
                                    ClusterService clusterService, IndicesService indicesService, ThreadPool threadPool,
                                    ShardStateAction shardStateAction, ActionFilters actionFilters, Writeable.Reader<Request> request,
-                                   Writeable.Reader<ReplicaRequest> replicaRequest, String executor, boolean forceExecutionOnPrimary) {
+                                   Writeable.Reader<ReplicaRequest> replicaRequest, String executor, boolean forceExecutionOnPrimary,
+                                   WriteMemoryLimits writeMemoryLimits) {
+        // We pass ThreadPool.Names.SAME to the super class as we control the dispatching to the
+        // ThreadPool.Names.WRITE thread pool in this class.
         super(settings, actionName, transportService, clusterService, indicesService, threadPool, shardStateAction, actionFilters,
-              request, replicaRequest, executor, true, forceExecutionOnPrimary);
+            request, replicaRequest, ThreadPool.Names.SAME, true, forceExecutionOnPrimary);
+        this.executor = executor;
+        this.forceExecutionOnPrimary = forceExecutionOnPrimary;
+        this.writeMemoryLimits = writeMemoryLimits;
+    }
+
+    @Override
+    protected Releasable checkOperationLimits(Request request) {
+        if (shouldMarkCoordinatingBytes(request)) {
+            long operationSizeInBytes = primaryOperationSize(request);
+            return writeMemoryLimits.markCoordinatingOperationStarted(operationSizeInBytes);
+        } else {
+            return () -> {};
+        }
+    }
+
+    protected boolean shouldMarkCoordinatingBytes(Request request) {
+        return true;
+    }
+
+    @Override
+    protected Releasable checkPrimaryLimits(Request request) {
+        return writeMemoryLimits.markPrimaryOperationStarted(primaryOperationSize(request));
```

Comment:
I think we need to also check if we already accounted for the bytes here, since the request would normally arrive in the operation phase and then afterwards the primary phase is executed locally?

Reply:
This PR accounts coordinating and primary work under different metrics.